sympy/core/exprtools.py

```diff
@@ -1414,156 +1290,549 @@ def factor_nc(expr):
         return expr.func(*[factor_nc(a) for a in expr.args])
     expr = expr.func(*[expand_power_exp(i) for i in expr.args])
 
-    from sympy.polys.polytools import gcd, factor
-    expr, rep, nc_symbols = _mask_nc(expr)
+    return _NCPoly(expr).factor_one()
 
-    if rep:
-        return factor(expr).subs(rep)
-    else:
-        args = [a.args_cnc() for a in Add.make_args(expr)]
-        c = g = l = r = S.One
-        hit = False
-        # find any commutative gcd term
-        for i, a in enumerate(args):
-            if i == 0:
-                c = Mul._from_args(a[0])
-            elif a[0]:
-                c = gcd(c, Mul._from_args(a[0]))
-            else:
-                c = S.One
-        if c is not S.One:
-            hit = True
-            c, g = c.as_coeff_Mul()
-            if g is not S.One:
-                for i, (cc, _) in enumerate(args):
-                    cc = list(Mul.make_args(Mul._from_args(list(cc))/g))
-                    args[i][0] = cc
-            for i, (cc, _) in enumerate(args):
-                if cc:
-                    cc[0] = cc[0]/c
-                else:
-                    cc = [1/c]
-                args[i][0] = cc
-        # find any noncommutative common prefix
-        for i, a in enumerate(args):
-            if i == 0:
-                n = a[1][:]
+
+def _nc_generators(expr):
+    """Return noncommutative polynomial generators of an expression.
+
+    Commutative symbols are excluded, and functions of generators other than
+    positive integer powers are treated as generators, as with the polynomials module
+    """
+    if expr.is_Add or expr.is_Mul:
+        generators = set()
+        for arg in expr.args:
+            generators |= set(_nc_generators(arg))
+        return generators
+    if expr.is_commutative:
+        return set()
+    if expr.is_Atom:
+        return set([expr])
+    if expr.is_Pow:
+        term, pw = expr.args
+        if pw.is_positive and pw.is_Integer:
+            return _nc_generators(term)
+    return [expr]
```

Comment:
Why not always return set? It looks like mistake
```suggestion
    return {expr}
```
I think that we can also get rid of unnecessary coercing `set(_nc_generators(arg))`